### ISSUE TITLE
Redis Grain Directory does not remove the key in Redis if the silo address was cleaned from the membership table

### ISSUE THREAD

**Issue by VladislavRybnikov** [dotnet/orleans#7066]
Hi,

We are using Orleans 3.3.0 with Redis grain directory and Consul membership table provider and observed some issues with actors trying to activate on the silo that was **removed**  from Consul after default cleanup period (7 days). 

Example of exception:
```
No target activation S10.207.149.18:11200:357113337*cli/4a504b8a@acf866bd for response message: Transient Rejection (info: Exception while sending message: Orleans.Runtime.Messaging.ConnectionFailedException: Unable to connect to endpoint S10.47.130.160:11200:355373853. See InnerException
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Orleans.Internal.OrleansTaskExtentions.MakeCancellable[T](Task`1 task, CancellationToken cancellationToken)
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   --- End of inner exception stack trace ---
   at Orleans.Runtime.Messaging.ConnectionManager.ConnectAsync(SiloAddress address)
   at Orleans.Runtime.Messaging.ConnectionManager.GetConnectionAsync(SiloAddress endpoint)
   at Orleans.Runtime.Messaging.OutboundMessageQueue.<SendMessage>g__SendAsync|10_0(ValueTask`1 c, Message m)) Response S10.47.130.160:11200:355373853*grn/16567B52/00000000+609334678@86b569bb->S10.207.149.18:11200:357113337*cli/4a504b8a@acf866bd #66830793
```

In our application, the case when the actor will be reactivated next time after a long period of time (several months) are very often and non-deterministic, also we have frequent deploys so this problem is very relevant for us.

I saw the discussion inside [issue 7054](https://github.com/dotnet/orleans/issues/7054) but didn't find any solution.
1. Do we really need to extend DefunctSiloCleanupPeriod to some long period (the year for example) to avoid such situations?
2. Maybe it is possible to handle such exceptions from the client code and clean up keys manually?
3. Also, maybe there is a possibility to fix this from the Orleans side? For example, change an algorithm of determining that the silo is known to be dead or handle connection errors by removing the key from Redis after multiple transient exceptions.

Thank you very much, and looking for your reply.

**Comment by benjaminpetit** [comment]
> Do we really need to extend DefunctSiloCleanupPeriod to some long period (the year for example) to avoid such situations?

It's indeed a possibility, but but careful the table might grow too much. You could also set the `RedisGrainDirectoryOptions.EntryExpiry`, but I strongly encourange you NOT to do it, since it may lead to duplicate activations, unless you are 100% sure that your activation will not live more than the value you set.

> Maybe it is possible to handle such exceptions from the client code and clean up keys manually?

Possible, but I don't think it's a good idea.

> Also, maybe there is a possibility to fix this from the Orleans side? For example, change an algorithm of determining that the silo is known to be dead or handle connection errors by removing the key from Redis after multiple transient exceptions.

One breaking change that we could do is store the membership table version in the redis entry: if the entry point to a silo that is not found in the membership table and the version stored in the entry is less than the version from the membership table, we can safely delete the entry.

I think I will implement that for 3.5.0

**Comment by VladislavRybnikov** [comment]
Thanks for reply! We will temporary tune our membership table provider clean up period and will be waiting for 3.5.0,cause the idea with storing membership table snapshot version in Redis sounds great!

**Comment by benjaminpetit** [comment]
Should be fixed in #7133
